### PR TITLE
fix: add timeout to FallbackProvider::call to unblock stalled providers

### PR DESCRIPTION
**Pull request by sophoah**
### Description

`FallbackProvider::call()` has no bound on how long it waits for a single provider to respond. If a provider TCP-connects but then never replies (as opposed to refusing the connection or erroring out), the call to it just hangs forever. Since the fallback loop awaits each provider in priority order, one stalled provider blocks the entire loop: no other provider in the list ever gets tried, no error is ever recorded, and nothing fires an alert because from the outside nothing looks like a failure, it looks like the call is still in flight.

We hit this in production: an overloaded RPC node kept its connection open without responding, and a validator wedged on it with no visible error.

Added a `call_timeout` field to `FallbackProvider` (default 30s, in line with other timeouts already used in this codebase such as `QUORUM_RPC_CALL_TIMEOUT` and `HTTP_CLIENT_TIMEOUT`), with a `with_call_timeout()` builder method for overriding it, mirroring the existing `max_block_time` / `with_max_block_time()` pattern.

The per-provider call in the fallback loop is now wrapped in `tokio::time::timeout`. A timeout produces a normal `ChainCommunicationError`, so it flows through the existing error handling unchanged, the provider's failure count increments and it gets deprioritized after enough failures, same as any other kind of failure.

`handle_stalled_provider` only runs when the provider actually responded to the main call, since retrying it after a timeout would just hang on the same stalled provider a second time. But that function makes its own call to check the provider's block height, and the first pass left that call unbounded: a provider that answers the main call fine but then hangs on this follow-up check could still block the loop, the same failure mode this PR set out to fix, one level down. Caught in review, that call is now wrapped in the same `call_timeout`, falling back to the last known block height on timeout the same way it already did on any other error.

### Drive-by changes

None.

### Related issues

None directly. #2189 covers a related but distinct failure mode (a provider that responds but reports a stale block height, handled by the existing `handle_stalled_provider` check) and was already closed by that work in 2023. This PR addresses a provider that never responds at all, which nothing currently bounds.

### Testing

Unit Tests.

Added `test_call_timeout_unblocks_stalled_provider`, which builds a `FallbackProvider` with a 50ms call timeout, has the first provider sleep past that timeout and the second provider respond normally, and asserts the call still returns `Ok` within a bounded wall clock deadline (wrappedin an outer timeout so the test itself fails loudly if the fix regresses) and that the stalled provider's failure count incremented.

Added `test_call_timeout_unblocks_stalled_block_height_check` for the follow-up fix: a provider responds fine to the main call but hangs on the block-height check, `max_block_time` is set to zero so that check isn't skipped as "too early to tell", and the assertion is the same, the call still returns within the outer timeout instead of hanging.

Ran `cargo fmt --check`, `cargo clippy -p hyperlane-core -- -D warnings`, and the full `fallback` test module (`cargo test -p hyperlane-core -p hyperlane-base --lib -- fallback`). Both the new test and the existing `test_deprioritization_by_failed_count` pass.

### CODE REVIEW

**Review by paulbalaji — COMMENTED** [review]
Reviewed exact head `5ebed5b829cc595569b0d309ef85711c55f9a1b3`. The focused fallback module passes (2/2).

I independently confirmed the existing unresolved staleness-probe finding: `tokio::time::timeout` bounds only `f(provider.clone())`; after a successful response, `handle_stalled_provider()` can still await `get_block_number()` without a deadline and wedge before the next fallback provider. No additional findings, but this is not ready until that existing thread is addressed.

Current GitHub checks are limited to four lightweight checks; the full test workflow has not run.

File: rust/main/hyperlane-core/src/rpc_clients/fallback.rs [hunk 1]
```diff
@@ -29,6 +29,9 @@ const MAX_BLOCK_TIME: Duration = Duration::from_secs(2 * 60);
 
 const FAILED_REQUEST_THRESHOLD: u32 = 10;
 
+// Caps how long we wait on one provider; otherwise a stalled connection blocks the whole loop.
+const FALLBACK_PROVIDER_CALL_TIMEOUT: Duration = Duration::from_secs(30);
+
```
[the remark this answers is not in the document]
Comment (paulbalaji) [reply rust/main/hyperlane-core/src/rpc_clients/fallback.rs:34]: Confirmed. The required fix is to put the successful-response health check inside the same timed attempt, so the 30s budget covers both `f(provider.clone())` and `handle_stalled_provider()`/its `get_block_number()` call. A timeout during either phase should become the normal timeout error and continue to the next provider. `handle_failed_provider()` is local lock/bookkeeping only, so I would leave it outside.

Please also extend the test with `with_max_block_time(Duration::ZERO)` and a `BlockNumberGetter` that hangs, then assert the second provider is reached. That directly catches the remaining wedge.

File: rust/main/hyperlane-core/src/rpc_clients/fallback.rs [hunk 2]
```diff
@@ -173,10 +178,11 @@ where
...
         let block_getter: B = provider.clone().into();
-        let current_block_height = block_getter
-            .get_block_number()
-            .await
-            .unwrap_or(priority.last_block_height.0);
+        let current_block_height =
+            tokio::time::timeout(self.call_timeout, block_getter.get_block_number())
+                .await
+                .unwrap_or(Ok(priority.last_block_height.0))
+                .unwrap_or(priority.last_block_height.0);
```
[the remark this answers is not in the document]
Comment (paulbalaji) [reply rust/main/hyperlane-core/src/rpc_clients/fallback.rs]: Confirmed — this is the remaining issue. The nested timeout bounds the hang, but it creates a second full 30s window and then returns the first provider’s response even though its block-height health check timed out. Please wrap the request plus successful-response health check in one timed attempt; a timeout in either phase should increment the provider failure and continue to the next provider. The regression test should use a second provider and assert it is reached.

[the remark this answers is not in the document]
Comment (sophoah) [reply rust/main/hyperlane-core/src/rpc_clients/fallback.rs]: addressed :
> Make the request and the successful-response health check one timed attempt.
Route an attempt timeout through the existing failure path.
Update the regression test to verify fallback to a second provider.

kept:
> Remove the timeout from handle_stalled_provider().

because hyperlane-ethereum and hyperlane-starknet call handle_stalled_provider() directly, bypassing call(), and it's their only bound on the probe. Removing it reopens this exact bug for Ethereum's real RPC path.

pushed commit ef7637c1c180a6ae627f3f0d51a7173083b44561

**Review by paulbalaji — APPROVED** [review]
Rereviewed exact head ef7637c1. The prior blocker is fixed: the provider request and successful-response health probe share one timeout budget, timeout enters normal failure accounting, and the regression proves fallback reaches provider 2. Focused fallback tests pass (3/3), and hyperlane-ethereum + hyperlane-starknet compile cleanly. No remaining findings.
